### PR TITLE
Added EQCCT PT model and tests

### PR DESCRIPTION
**Pull request by Costaki33**
This PR is looking to add the EQCCT phase picking model to SeisBench.
We added:

- EQCCTP (P-branch) and EQCCTS (S-branch) as WaveformModel subclasses
- Included unit tests in tests/models/test_eqcct.py, with all tests passing
- Provided model metadata in contrib/eqcct and contrib/eqccts
- Updates .gitignore to exclude model weights (*.pt)

We have the pretrained weights for the P and S models, please let me know where I should upload them so they can be added to the SeisBench data repository.

On an important note, EQCCT was originally created and published as a TensorFlow model. I recently created a pipeline to transfer the weights to PyTorch. That work is currently under publication review. I understand that you require a publication for contribution to SeisBench, please let me know if you would like to read my preprint so you can have the proper notes needed to discern if EQCCT can be contribute to SeisBench before or after that work is published.

Additionally, I forgot to add a `model extension` label to this PR, could someone with write access please add it so that the PR is standardized?

Thank you so much.

Best,
Constantinos

### CODE REVIEW

**Comment by yetinam** [comment]
Hi @Costaki33 ,
thanks a lot for your contribution! I'm looking forward to integrating EQCCT into SeisBench. Regarding the publication, given that EQCCT has been published before and it's now just a reimplementation in a different framework, I'm fine adding it to SeisBench already. I've started the CI and will look at the PR in detail when I find some time, probably next week.

**Comment by Costaki33** [comment]
Great! Sounds good, thanks so much. Please keep me updated.

File: contrib/eqcct/original.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  "docstring": "EQCCT P-branch — institutional PyTorch checkpoint in SeisBench format (6000 samples, 100 Hz, ZNE, peak norm). Pair with original.pt in ``contrib/eqcct/`` or ``<cache_model_root>/eqcct/`` after export.",
+  "model_args": {
+    "sampling_rate": 100,
+    "norm": "peak",
+    "norm_amp_per_comp": false,
+    "norm_detrend": false
+  },
+  "default_args": {
+    "batch_size": 256,
+    "overlap": 3000,
```
Comment (yetinam) [inline contrib/eqcct/original.json]: This is correct, but I would suggest using `0.5` instead. It was introduced in a later SeisBench version, but I think it makes the intent a little more clear.

File: contrib/eqcct/original.json [hunk 2]
```diff
@@ -0,0 +1,16 @@
+{
+  "version": "1",
+  "docstring": "EQCCT P-branch — institutional PyTorch checkpoint in SeisBench format (6000 samples, 100 Hz, ZNE, peak norm). Pair with original.pt in ``contrib/eqcct/`` or ``<cache_model_root>/eqcct/`` after export.",
```
Comment (yetinam) [inline contrib/eqcct/original.json]: Can you add a reference to the paper to the docstring?

File: seisbench/models/eqcct.py [hunk 3]
```diff
@@ -0,0 +1,554 @@
...
+EQCCT uses separate models for P and S; load with
+EQCCTP.from_pretrained(...) and EQCCTS.from_pretrained(...) respectively.
+Weights live under <cache_model_root>/eqcct/ and …/eqccts/ (not eqcctp/), or
+under contrib/eqcct / contrib/eqccts in a source checkout when both original.pt and
+original.json are present.
+"""
+
+from __future__ import annotations
+
+import math
```
Comment (yetinam) [inline seisbench/models/eqcct.py]: For consistency with other code, can you replace the usage with numpy?

File: seisbench/models/eqcct.py [hunk 4]
```diff
@@ -0,0 +1,554 @@
...
+    """
+    cdir = _contrib_weights_dir(cache_subdir)
+    if cdir is not None:
+        if (cdir / "original.pt").is_file() and (cdir / "original.json").is_file():
+            return cdir
+    return Path(seisbench.cache_model_root) / cache_subdir
+
+
+class ConvF1Block(nn.Module):
+    def __init__(self, in_channels, out_channels, kernal_size=11, dropout_rate=0.1):
```
Comment (yetinam) [inline seisbench/models/eqcct.py]: Typo `kernel_size`

File: seisbench/models/eqcct.py [hunk 5]
```diff
@@ -0,0 +1,554 @@
...
+
+class PatchEncoder(nn.Module):
+    def __init__(self, num_patches, projection_dim, patch_dim):
+        super().__init__()
+        self.projection = nn.Linear(patch_dim, projection_dim)
+        self.position_embedding = nn.Embedding(num_patches, projection_dim)
+
+    def forward(self, x):
+        positions = (
+            torch.arange(x.size(1), device=x.device).unsqueeze(0).expand(x.size(0), -1)
```
Comment (yetinam) [inline seisbench/models/eqcct.py]: It might be good to explicitly specify `dtype=x.dtype`

File: seisbench/models/eqcct.py [hunk 6]
```diff
@@ -0,0 +1,554 @@
...
+    :param norm: \"peak\" or \"std\" for amplitude normalization (default \"peak\").
+    :param kwargs: Passed to :class:`~seisbench.models.base.WaveformModel`.
+    """
+
+    @classmethod
+    def _model_path(cls):
+        return _eqcct_weights_directory("eqcct")
+
+    def __init__(self, sampling_rate=100, norm="peak", **kwargs):
+        citation = (
```
Comment (yetinam) [inline seisbench/models/eqcct.py]: This should reference a paper explicitly. I'd suggest referencing the original EQCCT publication for now. You can still switch it to the PyTorch publication later with a quick PR if you'd like.

**Comment by yetinam** [comment]
Hi @Costaki33 ,

thanks again for your contribution! It's of great quality! I've looked through it and only have few comments. There's a couple of detail comments that I put in the code. There are two slightly larger points.

I've seen that you did some work to reset the weights directory to the contrib folder. While I appreciate the effort, I'm not sure this is necessary. I'd be happy to upload the weights to the repository and then we can rely on the regular SeisBench infrastructure. That's what we've done for other model contributions. Can you send me the model weights (or a link to them) by mail? Then I'll upload them.

Could you add the model to the documentation? It should be added [here](https://seisbench.readthedocs.io/en/stable/pages/models/waveform_models.html) and [here](https://seisbench.readthedocs.io/en/stable/pages/documentation/models/waveform_models.html). I think it's enough to only add the main classes to the API reference, instead of all auxiliary classes used by the model (similar to how it's done for EQTransformer).

Once again, thanks for your contribution and let me know if you have any questions.

**Comment by Costaki33** [comment]
Hi @yetinam,

Thanks for these comments, I really appreciated them.

I've gone ahead and pushed your requested updates. Please let me know if the citations are fine, and if any other changes are needed.

I'll go ahead and email you the model weights so you can upload them. I've gone ahead and also removed the work I did to reset the weights directory in the contrib folder. That was to make sure I had something working and stable, but now that's not needed.

Thanks again for your help.

**Comment by yetinam** [comment]
Hi @Costaki33 ,
sorry for the long wait. I've now finished reviewing the new version of the PR. From my side, it's ready for merging. I made a few small changes, so I'd briefly ask you to confirm that they are okay before I merge the PR.

The biggest change is that I've switched the normalization from `peak` to `std`. The performance is much better and from what I see, this is the default for the [tensorflow version](https://github.com/ut-beg-texnet/eqcct/blob/1ea47724ff39ca0300d465afadf0877b46d3c7d5/eqcctone/eqcct/stream.py#L459). Can you confirm this?

I removed the json files from config. Instead, they are now [in the SeisBench repo](https://hifis-storage.desy.de/Helmholtz/HelmholtzAI/SeisBench/models/v3/), together with the weights.

All other changes are purely cosmetic. Let me know if this is okay with you. Thanks again for the PR!

**Comment by Costaki33** [comment]
Hi @yetinam,

No worries at all. And yes, these changes look great and I agree it's ready for merging.

Thanks again for helping me with this project, I really appreciate it.

Let me know if I can be of any more assistance.

**Comment by yetinam** [comment]
I just merged the PR. EQCCT will be part of the v0.12 release. There are a few other features in the pipeline right now, but I'm aiming to put out this release later in the summer. Thanks again for your contribution, @Costaki33!
